File: src/main/java/com/baulsupp/oksocial/credentials/PreferencesCredentialsStore.java
package com.baulsupp.oksocial.credentials;

import java.util.Optional;
import java.util.prefs.Preferences;

public class PreferencesCredentialsStore implements CredentialsStore {
  private final Optional<String> tokenSet;
  private Preferences userNode = Preferences.userNodeForPackage(this.getClass());

  public PreferencesCredentialsStore(Optional<String> tokenSet) {
    this.tokenSet = tokenSet;
  }

  @Override public <T> Optional<T> readDefaultCredentials(ServiceDefinition<T> serviceDefinition) {
    String credentialsString = userNode.get(tokenKey(serviceDefinition.apiHost()), null);
    return Optional.ofNullable(credentialsString)
        .map(serviceDefinition::parseCredentialsString);
  }

  private String tokenKey(String name) {
    return name + ".token" + tokenSet.map(s -> "." + s).orElse("");
  }

  @Override
  public <T> void storeCredentials(T credentials, ServiceDefinition<T> serviceDefinition) {
    String credentialsString = serviceDefinition.formatCredentialsString(credentials);
    userNode.put(tokenKey(serviceDefinition.apiHost()), credentialsString);
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/credentials/PreferencesCredentialsStore.kt
package com.baulsupp.oksocial.credentials

import java.util.prefs.Preferences

class PreferencesCredentialsStore : CredentialsStore {
  private val userNode = Preferences.userNodeForPackage(this.javaClass)

  override fun <T> get(serviceDefinition: ServiceDefinition<T>, tokenSet: String): T? {
    val credentialsString = userNode.get(tokenKey(serviceDefinition.apiHost(), tokenSet), null)
    return credentialsString?.let { serviceDefinition.parseCredentialsString(it) }
  }

  private fun tokenKey(name: String, tokenSet: String): String {
    return "$name.token.$tokenSet"
  }

  override fun <T> set(
    serviceDefinition: ServiceDefinition<T>, tokenSet: String, credentials: T) {
    val credentialsString = serviceDefinition.formatCredentialsString(credentials)
    userNode.put(tokenKey(serviceDefinition.apiHost(), tokenSet), credentialsString)
  }

  override fun <T> remove(serviceDefinition: ServiceDefinition<T>, tokenSet: String) {
    userNode.remove(tokenKey(serviceDefinition.apiHost(), tokenSet))
  }
}
